optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -53,31 +62,28 @@
      * Maximum capacity: The maximum capacity for each resource for each machine must not be exceeded.
      */
     private Constraint maximumCapacity(ConstraintFactory factory) {
-        throw new UnsupportedOperationException("Not yet implemented due to missing support for bi-grouping.");
-//        return factory.from(MrMachineCapacity.class)
-//                .join(MrProcessAssignment.class,
-//                        equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
-//                )
-//                .groupBy(
-//                        (machineCapacity, processAssignment) -> machineCapacity, sumLong(
-//                                (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource())
-//                        )
-//                )
-//                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
-//                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
-//                        HardSoftLongScore.ofHard(1L),
-//                        (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
+        return factory.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
+                )
+                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
+                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
+                              HardSoftLongScore.ofHard(1L),
+                              (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
     }
 
     /**
      * Conflict: Processes of the same service must run on distinct machines.
      */
     private Constraint serviceConflict(ConstraintFactory factory) {
         return factory.fromUniquePair(MrProcessAssignment.class,
-                equal(MrProcessAssignment::getMachine),
-                equal(MrProcessAssignment::getService)
+                                      equal(MrProcessAssignment::getMachine),
+                                      equal(MrProcessAssignment::getService)
         ).penalize(MrConstraints.SERVICE_CONFLICT,
-                HardSoftLongScore.ofHard(1L));
+                   HardSoftLongScore.ofHard(1L));
```

Comment:
FYI All instances of `HardSoftLongScore.ofHard(1L)` may be replaced by `HardSoftLongScore.ONE_HARD`,

Reply:
(Similarly for `ONE_SOFT`.)